File: leetcode/128.go
func longestConsecutive(nums []int) int {
    us := map[int]bool{}
    for _, num := range nums {
        us[num] = true
    }
    
    var ans int = 0
    
    for key, _ := range us {
        if _, exist := us[key-1]; !exist {
            startTarget := key
            currentLen := 1
            
            for  {
                _, ok := us[startTarget+1]
                if !ok {
                    break
                }
                startTarget ++
                currentLen ++
            }
            
            ans = max(ans, currentLen);
        }
    }
    return ans
}

func max(a, b int) int {
    if a > b {
        return a
    }
    return b
}

File: leetcode/128.cpp
class Solution {
public:
    int longestConsecutive(vector<int>& nums) {
        unordered_set<int> us(nums.begin(), nums.end());
        
        int ans = 0;
        for (int num : us) {
            // if num-1 is in the set, it means that num is part of some consecutive sequence. We don't need to calculate the sub-sequence!!
            if (us.find(num-1) == us.end()) {
                int startTarget = num;
                int currentLen = 1;

                while (us.find(startTarget+1) != us.end()) {
                    startTarget += 1; 
                    currentLen += 1;
                }

                ans = max(ans, currentLen);
            }
        }
        return ans;
    }
};
